nixos/borgbackup: change --prefix to --glob-archives

bjornfor opened the pull request:
###### Description of changes

Fixes warning from `borg prune`:

```
  Warning: "--prefix" has been deprecated. Use "--glob-archives 'yourprefix*'" (-a) instead.
```

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

bjornfor committed:
nixos/borgbackup: change --prefix to --glob-archives

Fixes warning from `borg prune`:

  Warning: "--prefix" has been deprecated. Use "--glob-archives 'yourprefix*'" (-a) instead.

ncfavier commented on nixos/modules/services/backup/borgbackup.nix:
```diff
@@ -58,7 +58,7 @@ let
     borg prune $extraArgs \
       ${mkKeepArgs cfg} \
-      ${optionalString (cfg.prune.prefix != null) "--prefix ${escapeShellArg cfg.prune.prefix} \\"}
+      ${optionalString (cfg.prune.prefix != null) "--glob-archives ${escapeShellArg "${cfg.prune.prefix}*"} \\"}
```
```suggestion
      ${optionalString (cfg.prune.prefix != null) "--glob-archives ${escapeShellArg "${cfg.prune.prefix}*"}"} \
```

That way we don't fail when `prune.prefix` is `null` and `extraPruneArgs` is set.

bjornfor replied:
To be honest, I don't understand what happens at the end of that line, nor why your suggestion improves things. (And isn't prefix == null already handled?)

Are these changes you think need to be done as part of the "change --prefix to --glob-archives" commit or improvements you'd like to have, that can be done as follow up?

Sorry, I don't have brain cycles left... unless this causes regressions I suggest to land as-is, and people can make improvements on top of master?

ncfavier replied:
The precise situation where this is a problem is when `prune.prefix == null && extraPruneArgs != ""`.

With the backslash inside the optional string, the generated command looks like this:
```sh
echo borg create \
  
  extra args
```
and running it results in
```
borg create
script: line 3: extra: command not found
```

With the backslash on the outside, the generated command looks like this:
```sh
echo borg create \
  \
  extra args
```
and running it results in
```
borg create extra args
```

This doesn't have anything to do with your change, just an improvement. I'll take responsibility for it and just commit it and merge.

ncfavier committed:
nixos/borgbackup: fix newline escaping with optional arguments

The newline must always be escaped, otherwise `extraPruneArgs` will be treated as a separate command if `prune.prefix` is unset.

bjornfor commented:
@ncfavier: Thanks for explaining and commit+merge!
